### ISSUE TITLE
Match multiple parameters in an Object in their origin order?

### ISSUE THREAD

**Issue by roastduck** [cinemast/libjson-rpc-cpp#81]
Currently, when I specify parameters in an Object, that means not an Array, in spec file, the parser will deal with them in lexicographical order. For example, if in the spec file there is _"params":{"b":0,"a":0}_, and the method is _void(int b, int a)_, the "a" will be matched to "b" and the "b" will be matched to "a". 

I know, although, while we are using virtual method, we can not exactly find which is "a" and which is "b", but I think people intend to define their method parameters in the same order with their spec file. Am I right? How about dealing the parameters in their origin order in the spec file? This will bring lots of convenience.  I know this may be kind of troublesome because we can not simply parse the spec file as JSON then, but it is better, right?

Do you think this is a good idea or I am wrong? Thanks.

**Comment by cinemast** [comment]
Hi!

I am not sure if I got what you described here. If you use an object as params description, this means you are using the named parameter scheme of JSON-RPC, in which case no order is specified, but the name of the parameters are used in the method signature.

If you would like to have positional parameters, change the params field to an array instead of an object.

Does this answer your question?

**Comment by roastduck** [comment]
Hi.

No, that isn't my question.

Supposed we have such a stubServer class declaration.

    class MyStubServer : public AbstractStubServer
    {
        public:
            MyStubServer(AbstractServerConnector &connector, serverVersion_t type);

            virtual int foo(int b, int a);
    };

Notice that _b_ is in front of _a_.

And we have such a _spec.json_ below.

    [
        "name": "foo",
        "params":
        {
            "b": 0,
            "a": 0
        },
        "returns": 0
    ]

Now we want to pass parameters as _"{"b":1, "a":2}"_, but actually we got _a=1_ and _b=2_. Because when we parse the _spec.json_, we consider the _"params"_ as a _JSON object_, which means we treat the elements inside not in order (actually we treat them in lexicographical order with out implementation). In the mean time, we define our method _foo_ as a virtual member function _int foo(int,int)_, which means C++ just recognize the type but not the parameter name, so the _a_ is matched into _b_ and the _b_ is matched into _a_.

As JSON_RPC described, when we use an _JSON object_ to pass parameters, we don't need to take the order of them into consideration. However, when we use this stub server, we must put the parameters in lexicographical order in the method definition, both in the class and the spec file, or it doesn't work. So here's the problem. Maybe we are not supposed to treat the spes file as JSON format?

**Comment by cinemast** [comment]
Ah, now I got your problem, thanks for describing it again.

Well there are 2 problems here, neither of them can be handled by me or this library. 

1.) C++ does not verify the parameter names of implemented abstract methods. If you look at the abstractstubserver.h the parameters are named correctly there (even though not in the "right" order).
But since you decided to use named parameters instead of positional parameters, it is up to you ensure that you implement the abstract method correctly.

2.) libjsoncpp, which is working under the hood, is responsible for parsing the JSON specification file. Somehow they decided to iterate over them lexicographically, which is ok, because the JSON standard defines no order in case of objects.

So in short: Yes, this is a source of potential errors, but there is nothing I can do about, except of changing the format of the specification file from JSON to something else. But I won't do that since this would break all existing applications build on top of libjson-rpc-cpp.

Maybe I overlooked something and you have a better idea (which is in fact implementable). In that case please let me know.

**Comment by roastduck** [comment]
I see that. Neither changing the way of parsing virtual function or JSON string is implementable, but how about pre-process the spec file before libjsoncpp read it?

We can write a pre-processer, which scan over the spec file, find the parameter part, and add some suffix to the parameter name, for example, _{"b":0, "a":0}_ now goes into _{"__1_b":0, "__2_a":0}_ and it is now in lexicographical order. Then, when we construct the true class, which inherit the class users defined, we check the parameter out and delete the suffix we added.

This sounds like a hack, but it is actually implementable and it won't put our code into a mess, I think.

**Comment by cinemast** [comment]
That would be possible, but I am not willing to implement that. I think it makes the whole process unnecessarily complex and the gain of functionality would be close to zero.

I think that most developers using this framework are aware of the way C++ implements abstract methods. The parameters are correctly named in the abstract version of it. 

If you really require this, feel free to fork this framework. 

Hope you understand my point of view. If so, please close the issue, otherwise let me know.
